### PR TITLE
ENT-4681: Test proper scoping of classes defined from non-default namespace

### PR DESCRIPTION
**Pull request by nickanderson**

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thanks for submitting a PR! Maybe @vpodzime can review this?

File: tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+}
+body file control {  namespace => "test"; }
+bundle agent example
+{
+  classes:
+    "TRIGGERED"
+      expression => "any",
+      scope => "bundle"; # Bundle scoped classes are NOT visible from other bundles
+}
+body file control {  namespace => "default"; }
```
Comment (vpodzime) [inline tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf]: What is this double `body file control` supposed to do? And what makes you think it should be like that?

File: tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      meta => { "ENT-4681" };
+
+  methods:
+    "test:example";
+}
+bundle agent check 
+{
+  classes:
+    # Expectation, default:TRIGGERED is NOT defined
+    "defualt_TRIGGERED_OK"
```
Comment (vpodzime) [inline tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf]: typo here

File: tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-namespace-perspective.cf [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  classes:
+    "TRIGGERED"
+      expression => "any",
+      scope => "bundle"; # Namespace scoped classes are visible from other bundles
+}
+bundle agent check 
+{
+  classes:
+    # Expectation, default:TRIGGERED is NOT defined
+    "defualt_TRIGGERED_OK"
```
Comment (vpodzime) [inline tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-namespace-perspective.cf]: Same typo here and in all the other tests below.

File: tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf (hunk 1 above)
Comment (nickanderson) [reply tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf]: The first one set the namespace to test, the second one returned the namespace to default. So `bundle agent example` is in the `test` namespace but `bundle agent test` is in the default namespace.

File: tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf (hunk 2 above)
Comment (nickanderson) [reply tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf]: ACK

File: tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf (hunk 1 above)
Comment (nickanderson) [reply tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf]: > Please fix the typos. Also, is there a ticket for this issue?

ACK, and yes, ENT-4681 is the related ticket

Comment (vpodzime) [reply tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf]: I am quite surprised that the double `body file control` just works like this, sequentially. That looks almost like a crime! :wink:

Comment (olehermanse) [reply tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf]: I agree with @vpodzime this looks insane to me. The ordering of bodies and bundles within a file shouldn't matter, at all. Also a `body file control` is the control body for the file no? It should be for the whole file. I would expect the last body to overwrite the attributes of the first one, or an error.

Comment (nickanderson) [reply tests/acceptance/00_basics/04_bundles/agent-bundle-class-bundle-scope-default-perspective.cf]: Yeah, but this is how it was designed.

https://docs.cfengine.com/docs/3.12/reference-components-file_control_promises.html

```
This directive can be given multiple times within any file, outside of body and bundle definitions.
```

**Comment by olehermanse** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2745/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2745/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2745/

**Pull request merged**